### PR TITLE
[RNMobile] Embed block: Enable replace option

### PR DESCRIPTION
**Pull request by fluiddot**
`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3703

## Description

Implement and enable the replace option as described in [this issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3285).

### Replace URL
1. Add an Embed block
2. Set a URL
3. Dismiss the bottom sheet either by tapping outside or tapping the Intro keyboard button
4. Tap on ✏️ button located in the toolbar to replace the URL
5. Set a different URL
6. Dismiss the bottom sheet either by tapping outside or tapping the Intro keyboard button
7. Observe that the URL has been replaced and preview displays the new URL

### Invalid URL
1. Add an Embed block
2. Set a URL
3. Dismiss the bottom sheet either by tapping outside or tapping the Intro keyboard button
4. Tap on ✏️ button located in the toolbar to replace the URL
5. Set an invalid URL (example: `http://`)
6. Dismiss the bottom sheet either by tapping outside or tapping the Intro keyboard button
7. Observe that the URL hasn't been replaced and preview displays the previous URL

### Empty URL
1. Add an Embed block
2. Set a URL
3. Dismiss the bottom sheet either by tapping outside or tapping the Intro keyboard button
4. Tap on ✏️ button located in the toolbar to replace the URL
5. Set an empty URL
6. Dismiss the bottom sheet either by tapping outside or tapping the Intro keyboard button
7. Observe that the URL has been replaced with an empty value and the block's empty state is displayed

## Screenshots 

https://user-images.githubusercontent.com/14905380/124735898-c27a4680-df16-11eb-9ff2-c8016db7c80c.mp4

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/embed/edit.js [hunk 1]
```diff
@@ -238,6 +238,7 @@ const EmbedEdit = ( props ) => {
 						invalidateResolution( 'getEmbedPreview', [ url ] );
 					} }
 					isSelected={ isSelected }
+					isEditingURL={ isEditingURL }
```
Comment (fluiddot) [inline packages/block-library/src/embed/edit.js:241]: `isEditingURL` is `true` when the replace action is triggered.

File: packages/block-library/src/embed/embed-placeholder.native.js [hunk 2]
```diff
@@ -39,7 +40,7 @@ const EmbedPlaceholder = ( {
 		[ clientId ]
 	);
 	const [ isEmbedSheetVisible, setIsEmbedSheetVisible ] = useState(
-		isSelected && wasBlockJustInserted && ! value
+		isSelected && ( ( wasBlockJustInserted && ! value ) || isEditingURL )
```
Comment (fluiddot) [inline packages/block-library/src/embed/embed-placeholder.native.js:43]: This change enables the bottom sheet to be automatically displayed when the replace action is triggered.

**Review by ceyhun — CHANGES_REQUESTED** [review]
Thanks Carlos! Replacing and emptying URL works great, but I couldn't trigger invalid error case neither on iOS nor on Android. But I think it's not related to the changes in this PR.  `isURL` seems to return `true` for `'http://test'`, but `false` for just `'test'`.

**Comment by fluiddot** [comment]
> Thanks Carlos! Replacing and emptying URL works great, but I couldn't trigger invalid error case neither on iOS nor on Android. But I think it's not related to the changes in this PR. `isURL` seems to return `true` for `'http://test'`, but `false` for just `'test'`.
> 
> 

I'm not sure if I understood what you commented, do you mean that we should display the invalid error message for the value `test`?

I'd like to note that `LinkSettings`, the component we use for the URL text field, automatically prepends the `http` part ([code reference](https://github.com/WordPress/gutenberg/blob/1eaad3cbad327e9fad5d8eef93c46aa773b7080b/packages/components/src/mobile/link-settings/index.native.js#L140-L144)) if the URL doesn't contain a protocol, so there are just a few cases where with this logic, we would show the invalid error message.

**Comment by ceyhun** [comment]
> I'm not sure if I understood what you commented, do you mean that we should display the invalid error message for the value test?

I think `http://test` or `test` shouldn't be a valid URL. Otherwise I'm not sure if there is another case that can trigger the error message here:

https://github.com/WordPress/gutenberg/blob/a9798cc088778c2f91ef91d3c6e21deeb579370c/packages/block-library/src/embed/embed-bottom-sheet.native.js#L41-L43

> I'd like to note that LinkSettings, the component we use for the URL text field, automatically prepends the http part (code reference) if the URL doesn't contain a protocol, so there are just a few cases where with this logic, we would show the invalid error message.

I think this was working before, because maybe we were using `onClose` to get the value in the bottom sheet and maybe that was returning the value without http(s)? Not sure what's better now, to strip the `http://` before checking with `isURL` or going back to using `onClose` or maybe something else?

**Comment by fluiddot** [comment]
I think there are two different validations on this block:
1. **URL format:** It just checks if the URL is not malformed, although it may look weird, `http://test` URL is valid 😄 . In fact, I used it in the web version and I didn't get any error message, however, the block was transformed to a Paragraph block probably because it detected that is not embeddable.
2. **URL is embeddable:**: It checks if the URL can be used for embedding content, in this case, `http://test` would fail.

> I think this was working before, because maybe we were using onClose to get the value in the bottom sheet and maybe that was returning the value without http(s)? Not sure what's better now, to strip the http:// before checking with isURL or going back to using onClose or maybe something else?

Not sure, from the code, looks like the prepend happens before calling the `setAttributes` callback, so I think it was happening before:
https://github.com/WordPress/gutenberg/blob/1eaad3cbad327e9fad5d8eef93c46aa773b7080b/packages/components/src/mobile/link-settings/index.native.js#L140-L144
https://github.com/WordPress/gutenberg/blob/1eaad3cbad327e9fad5d8eef93c46aa773b7080b/packages/components/src/mobile/link-settings/index.native.js#L162-L173

**Comment by ceyhun** [comment]
> URL format: It just checks if the URL is not malformed, although it may look weird, http://test URL is valid 😄 . In fact, I used it in the web version and I didn't get any error message, however, the block was transformed to a Paragraph block probably because it detected that is not embeddable.

Ah, I guess then `http://test` must be valid indeed 😄 I just tested with something like `///invalid` instead and I can get the error message, sorry for the trouble. I was able to check and confirm that the invalid case works as well 🎉

**Review by ceyhun — APPROVED** [review]

**Comment by fluiddot** [comment]
> > URL format: It just checks if the URL is not malformed, although it may look weird, http://test URL is valid 😄 . In fact, I used it in the web version and I didn't get any error message, however, the block was transformed to a Paragraph block probably because it detected that is not embeddable.
> 
> Ah, I guess then `http://test` must be valid indeed 😄 I just tested with something like `///invalid` instead and I can get the error message, sorry for the trouble. I was able to check and confirm that the invalid case works as well 🎉

No worries, I was also surprised when I saw that weird URLs like `http://test` were valid 😄.
Thank you so much for the review ❤️ !

**Pull request merged**